File: BST.hs
import qualified Data.List
------------Constructor----------------------
data Tree a = EmptyTree | Node a (Tree a) (Tree a)
        deriving (Show, Eq)
------------Insert---------------------------
insert :: (Ord a) => a -> Tree a -> Tree a
insert x EmptyTree = Node x EmptyTree EmptyTree
insert x (Node a left right)
        | x == a = Node x left right
        | x < a  = Node a (insert x left) right
        | x > a  = Node a left (insert x right)
------------Remove---------------------------
remove :: (Ord a) => a -> Tree a -> Tree a
remove _ EmptyTree = EmptyTree
remove delKey (Node dataI t1 t2)
        | delKey == dataI = removeHelper (Node dataI t1 t2)
        | delKey > dataI = Node dataI t1 (remove delKey t2)
        | delKey < dataI = Node dataI (remove delKey t1) t2
------------Remove Helper--------------------
removeHelper :: (Ord a) => Tree a -> Tree a
removeHelper (Node a EmptyTree t2) = t2 -- No left child
removeHelper (Node a t1 EmptyTree) = t1 -- No right child
removeHelper (Node a t1 t2) = (Node a2 t1 t2)
    where a2 = rightMost t1 --gets predessor
------------Left Most-----------------------
leftMost :: (Ord a) => Tree a -> a
leftMost (Node dada EmptyTree _) = dada
leftMost (Node _ t1 _) = leftMost t1
------------Right Most----------------------
rightMost :: (Ord a) => Tree a -> a
rightMost (Node dada _ EmptyTree) = dada
rightMost (Node _ _ t2) = rightMost t2
------------Retrieve-------------------------
retrieve :: (Ord a) => a -> Tree a -> Bool
retrieve x EmptyTree = False
retrieve x (Node a left right)
        | x == a = True
        | x < a  = retrieve x left
        | x > a  = retrieve x right
-----------List To Tree-------------------------
listToTree :: (Ord a) => [a] -> Tree a
listToTree [] = EmptyTree
listToTree (x:xs) = createTree (Node x EmptyTree EmptyTree) xs
    where
        createTree tree [] = tree
        createTree tree (x:xs) = createTree (insert x tree) xs
-----------isEmpty-----------------------------
empty :: (Ord a) => Tree a -> Bool
empty EmptyTree = True
empty _ = False
-----------PreOrder Print-----------------------
preorder :: (Ord a) => Tree a -> [a]
preorder EmptyTree = []
preorder (Node v t1 t2) = [v] ++ preorder t1 ++ preorder t2
-----------InOrder Print-----------------------
inorder :: (Ord a) => Tree a -> [a]
inorder EmptyTree = []
inorder (Node v t1 t2) = inorder t1 ++ [v] ++ inorder t2
-----------PostOrder Print-----------------------
postorder :: (Ord a) => Tree a -> [a]
postorder EmptyTree = []
postorder (Node v t1 t2) = postorder t1 ++ postorder t2 ++ [v]
-------------------------------------------------
-----------Show Structure------------------------
--showStructure :: Tree a -> IO ()
--showStructure (Node root t1 t2)= do
--        showHelper

File: RandomProjects/BST.cpp
#include <iostream>
#include "BST.h"

using namespace std;

BST::BST()
{
    root = NULL;
}

BST::BST(const BST& other)
{
    copyHelper(root, other.root);
}

BST& BST::operator=(const BST& other)
{
    if (this == &other)
        return *this;
    clear();
    copyHelper(root, other.root);
    return *this;
}

BST::~BST()
{
    clear();
}

void BST::insert(const int& newData)
{
    insertHelper(root, newData);
}

void BST::clear()
{
    clearHelper(root);
}

void BST::showStructure() const
{
    if ( root == 0 )
       cout << "Empty tree" << endl;
    else
    {
       cout << endl;
       showHelper(root,1);
       cout << endl;
    }
}

void BST::writeKeys() const
{
    writeKeysHelper(root);
}

void BST::writeKeysLessThan(const int& searchKey) const
{
    writeKeysLessThanHelper(root, searchKey);
}

bool BST::retrieve(const int& searchKey, int& newDataItem)
{
    return retrieveHelper(root, searchKey, newDataItem);
}

bool BST::remove(const int& searchKey)
{
    if (isEmpty())
        return false;
    else
        return removeHelper(root, searchKey);
}

bool BST::isEmpty() const
{
    return (root == NULL);
}

int BST::getHeight() const
{
    return getHeightHelper(root);
}

int BST::getCount() const
{
    return getCountHelper(root);
}

BST::Node::Node(const int& nodeDataItem, Node* leftPtr, Node* rightPtr)
{
    data = nodeDataItem;
    left = leftPtr;
    right = rightPtr;
}

// ======================= Helper functions ===================== //
//


void BST::showHelper(Node* p, int level) const
{
     int j;   // Loop counter

     if ( p != 0 )
     {
        showHelper(p->right,level+1);         // Output right subtree
        for ( j = 0 ; j < level ; j++ )    // Tab over to level
            cout << "\t";
        cout << " " << p->data;   // Output key
        if ( ( p->left != 0 ) &&           // Output "connector"
             ( p->right != 0 ) )
           cout << "<";
        else if ( p->right != 0 )
           cout << "/";
        else if ( p->left != 0 )
           cout << "\\";
        cout << endl;
        showHelper(p->left,level+1);          // Output left subtree
    }
}

void BST::copyHelper(Node*& p , Node* src)
{
    if (src == NULL)
        return;
    {
        insertHelper(p, src->data);
        copyHelper(p->left, src->left);
        copyHelper(p->right, src->right);
    }
}

void BST::clearHelper(Node*& p)
{
    if (p == NULL)
        return;
    clearHelper(p->left);
    clearHelper(p->right);
    delete p;
    p = NULL;
}

void BST::insertHelper(Node*& p, const int& searchKey)
{
    if (p == NULL)
    {
        p = new Node(searchKey, NULL, NULL);
        return;
    }
    if (p->data > searchKey)
        insertHelper(p->left, searchKey);
    else if (p->data < searchKey)
        insertHelper(p->right, searchKey);
    else
        p->data = searchKey;
}

void BST::writeKeysHelper(Node* p) const
{
    // insert code later
}

void BST::writeKeysLessThanHelper(Node* p, const int& searchKey) const
{
    // insert code later
}

int BST::getCountHelper(Node* p) const
{
    return 0;   // insert code later
}

int BST::getHeightHelper(Node* p) const
{
    return 0;   // insert code later
}

bool BST::removeHelper(Node*& p, const int& searchKey)
{
    if (p == NULL)
        return false;
    else if (p->data > searchKey)
        return removeHelper(p->left, searchKey);
    else if (p->data < searchKey)
        return removeHelper(p->right, searchKey);
    else
    {
        if (p->left == NULL && p->right == NULL)
        {
            delete p;
            p = NULL;
            return true;
        }
        if (p->right == NULL)
        {
            Node* temp = p;
            p = p->left;
            delete temp;
            return true;
        }
        if (p->left == NULL)
        {
            Node* temp = p;
            p = p->right;
            delete temp;
            return true;
        }
        Node* temp = p->left;
        while (temp->right != NULL)
            temp =  temp->right;
        p->data = temp->data;
        return removeHelper(p->left, temp->data);
    }
    return false;
}

bool BST::retrieveHelper(Node* p, const int& searchKey, int& searchDataItem) const
{
    if (p == NULL)
        return false;
    if (p->data == searchKey)
    {
        searchDataItem = p->data;
        return true;
    }
    if (p->data > searchKey)
        return retrieveHelper(p->left, searchKey, searchDataItem);
    else if (p->data < searchKey)
        return retrieveHelper(p->right, searchKey, searchDataItem);
    return false;
}


